### ISSUE TITLE
warn about round tripping through pointers

### ISSUE THREAD

**Issue by brooksdavis** [CTSRD-CHERI/llvm-project#409]
A significant number of CheriABI issues in CheriBSD have been caused by inappropriate casts of pointers through arbitrary integer types.  For example, this set of macros were fixed in CTSRD-CHERI/cheribsd@e3a69027cc5a384431156d61c90d4304387a9b9d

``` C
#define PTROF(X)       ((BUFHEAD *)((ptrdiff_t)(X)&~0x3))
#define ISMOD(X)       ((u_int32_t)(ptrdiff_t)(X)&0x1)
#define DOMOD(X)       ((X) = (char *)((ptrdiff_t)(X)|0x1))
#define ISDISK(X)      ((u_int32_t)(ptrdiff_t)(X)&0x2)
#define DODISK(X)      ((X) = (char *)((ptrdiff_t)(X)|0x2))
```

This case seems like it should be easy to detect and warn about during normal compiler operation.  Other cases where the casts in and out of an integer occur at some distance form each other may require a clang static analyzer module.

**Comment by eupharina** [comment]
I implemented this in CSA as a part of ProvenanceSourceChecker here: PR https://github.com/CTSRD-CHERI/llvm-project/pull/675.

Cast sequences that do not involve `(u)intptr_t` are not reported, I guess it's covered by the compiler warning.
Some examples of casts in and out of an integer at some distance (including inter-procedural) reported by checker can be found here: https://github.com/CTSRD-CHERI/llvm-project/pull/675#issuecomment-1384377264
